src/main/scala/devices/debug/Debug.scala

```diff
@@ -722,6 +716,184 @@ class TLDebugModuleInner(device: Device, getNComponents: () => Int, beatBytes: I
+       * SBADDRESS registers depends on system bus widths
+       */
+      val anyAddressWrEn = Wire(init = false.B)
+      val anyDataRdEn    = Wire(init = false.B)
+      val anyDataWrEn    = Wire(init = false.B)
+      anyAddressWrEn.suggestName("anyAddressWrEn")
+      anyDataRdEn.suggestName("anyDataRdEn")
+      anyDataWrEn.suggestName("anyDataWrEn")
+
+      // --- SBCS Status Register ---
+      val SBCSFieldsReg      = Reg(new SBCSFields())
+      val SBCSFieldsRegReset = Wire(init = (new SBCSFields()).fromBits(0.U))
+      val SBCSRdData         = Wire(init = new SBCSFields().fromBits(0.U)) 
+      val SBCSWrDataVal      = Wire(init = 0.U(32.W))
+      val SBCSWrData         = Wire(init = new SBCSFields().fromBits(SBCSWrDataVal))
+      val SBCSRdEn, SBCSWrEn = Wire(init = false.B)
+
+      val sbcsfields = Seq(RWNotify(32, SBCSRdData.asUInt(), SBCSWrDataVal, SBCSRdEn, SBCSWrEn,
+                          Some(RegFieldDesc("dmi_sbcs", "", reset=Some(0)))))
+
+      // --- System Bus Address Registers ---
+      // ADDR0 Register is required
+      // Instantiate ADDR1-3 registers as needed depending on system bus address width
+      val hasSBAddr1 = (sb2tl.module.edge.bundle.addressBits >= 33)
+      val hasSBAddr2 = (sb2tl.module.edge.bundle.addressBits >= 65)
+      val hasSBAddr3 = (sb2tl.module.edge.bundle.addressBits >= 97)
+      val hasAddr    = Seq(true, hasSBAddr1, hasSBAddr2, hasSBAddr3)
+
+      val SBADDRESSFieldsReg = Seq.fill(4)(Reg (UInt(32.W)))
+      val SBADDRESSRdData    = Seq.fill(4)(Reg (UInt(32.W)))
+      val SBADDRESSWrData    = Seq.fill(4)(Reg (UInt(32.W)))
+      val SBADDRESSRdEn      = Seq.fill(4)(Wire(Bool()))
+      val SBADDRESSWrEn      = Seq.fill(4)(Wire(Bool()))
+
+      val autoIncrementedAddr = Wire(init = 0.U(128.W))
+      autoIncrementedAddr := sb2tl.module.io.addrIn + (1.U << SBCSFieldsReg.sbaccess)
+
+      SBADDRESSFieldsReg.zipWithIndex.foreach { case(a,i) =>
+
+        if(hasAddr(i)) {
+          when (~io.dmactive) {
+            a := 0.U(32.W)
+          }.otherwise {
+            a := Mux(SBADDRESSWrEn(i) && !SBCSFieldsReg.sberror && !SBCSFieldsReg.sbbusy, SBADDRESSWrData(i),
+                 Mux(sb2tl.module.io.dataOutValid && SBCSFieldsReg.sbautoincrement,       autoIncrementedAddr(32*i+31,32*i), SBADDRESSFieldsReg(i)))
+          }
+
+          SBADDRESSRdData(i) := SBADDRESSFieldsReg(i)
+
+          sbaddrfields(i) = Seq(RWNotify(32, SBADDRESSRdData(i).asUInt(), SBADDRESSWrData(i), SBADDRESSRdEn(i), SBADDRESSWrEn(i),
+                            Some(RegFieldDesc("dmi_sbaddr"+i, "", reset=Some(0)))))
+        } else {
+          sbaddrfields(i) =  Seq.empty[RegField]
+        }
+      }
+
+      if(hasAddr(3)) {
+        sb2tl.module.io.addrIn := Cat(SBADDRESSRdData(3),SBADDRESSRdData(2),SBADDRESSRdData(1),SBADDRESSRdData(0))
+        anyAddressWrEn         := SBADDRESSWrEn(3) || SBADDRESSWrEn(2) || SBADDRESSWrEn(1) || SBADDRESSWrEn(0)
```

Comment:
You could simplify this a bit, by making `SBADDRESSWrEn = Seq.fill(4)(Wire (init = False.B))`, then `anyAddressWrEn := SBADDRESSWrEn.reduce(_ || _)` (no special casing needed).

Reply:
Thanks, I knew there'd be a way to do this better. Is there any similar trick for the address bus concatenation, or do I have to do special casing?

Reply:
I think if you try to assign `sb2tl.moduile.io.addrIn` to something that is widier than it, it will just truncate the lower bits anyway. So you can try `sb2tl.module.io.addrIn := Cat(SBADDRESSRdData.reverse:_*`